### ISSUE TITLE
CUDA default autocast (bf16) is the least precise dtype: up to 0.073 from fp32 and a preset answer flips; fp16 is within 0.019 at the same latency

### ISSUE THREAD

**Issue by bhushankinge** [NandhaKishorM/laya#443]
## What happened

`Agent.__init__` picks the CUDA autocast dtype from the checkpoint's `amp_dtype` whenever compute capability is >= 8 (`agent.py`, the "Autocast policy" block), and all three shipped checkpoints set `bf16`, so every Ampere-or-newer GPU runs the stock forward in bf16. Only `LAYA_CPU_AMP` exists as an override; there is no CUDA counterpart, and neither README nor BENCHMARKS says which dtype a CUDA user gets.

bf16 is the least precise of the dtypes the stock forward can run in. On `benchmarks/parity_fast.py`'s own fixed set (60 states x up to 8 questions, 288 questions per checkpoint), same tokens and same per-state batching as that script, raw softmax as it reports, fp32 forward as the reference:

| checkpoint | type | n | max \|p_bf16 - p_fp32\| | max \|p_fp16 - p_fp32\| | argmax bf16 = fp32 | fp16 = fp32 |
|---|---|---|---|---|---|---|
| laya | choice | 48 | 0.0185 | 0.0076 | 48/48 | 48/48 |
| laya | noul | 180 | **0.0731** | 0.0185 | 179/180 | 180/180 |
| laya | score | 60 | 0.0116 | 0.0052 | 60/60 | 60/60 |
| laya-multilingual | choice | 48 | 0.0391 | 0.0019 | 48/48 | 48/48 |
| laya-multilingual | noul | 180 | **0.0709** | 0.0068 | 180/180 | 180/180 |
| laya-multilingual | score | 60 | 0.0079 | 0.0021 | 60/60 | 60/60 |
| laya-typed-decisions | choice | 48 | 0.0124 | 0.0032 | 46/48 | 48/48 |
| laya-typed-decisions | noul | 180 | 0.0419 | 0.0045 | 180/180 | 180/180 |
| laya-typed-decisions | score | 60 | 0.0128 | 0.0013 | 60/60 | 60/60 |

fp16 stays within 0.019 of fp32 on every row and never changes an argmax; bf16 reaches 0.073 and flips three of 864. Median per-question error is 5-10x lower in fp16 (0.0001 vs 0.0005-0.0018). Latency is the same: `predict()`, one short state x 3 triage questions, 100 iterations after warm-up, p50 31.3 ms bf16 vs 31.2 ms fp16 (`laya`), 14.1 vs 13.5 (`multilingual`), 31.6 vs 32.2 (`typed-decisions`).

It reaches the public API. `laya` with the moderation preset on the parity set's outage text (`TEXTS[1]`, "Since yesterday nobody on our team can log in; the dashboard returns 502 ..."), one `predict()` call with the preset's questions, temperatures applied (noul:2 = 1.98):

| forward | spam `noul` | answer |
|---|---|---|
| fp32 (`amp_enabled=False`) | 0.468 | no |
| fp16 autocast | 0.469 | no |
| bf16 autocast (the default) | **0.505** | **yes** |

Ask the same spam question alone and bf16 also says no (0.4685): the bf16 error is large enough that the answer depends on which other questions share the padded batch. README already warns that batch shapes can move probabilities "particularly with mixed precision"; fp16 does not show it at this magnitude.

The same ordering held on four GPUs on a smaller 63-question set (RTX 2000 Ada, RTX PRO 5000, RTX PRO 6000, H100 NVL: bf16 0.004-0.016 vs fp16 0.0005-0.007, 63/63 argmax on all of them); the upstream set is just large enough to show the flips. The BENCHMARKS parity table already carries the signal: `max |p_stock - p_fp32|` for the stock bf16 path is 0.024 / 0.058 / 0.017 (`laya`) and 0.039 / 0.045 / 0.009 (`multilingual`) on an RTX 4070 Ti SUPER, the same order as the bf16 column above; it just is not compared against fp16 anywhere.

## Steps to reproduce

Every per-option probability, the script and the host details: https://github.com/bhushankinge/laya-cuda-bench/tree/main/results/laptop-rtx2000ada/dtype_parity

```bash
git clone https://github.com/bhushankinge/laya-cuda-bench && cd laya-cuda-bench
LAYA_SRC=<laya checkout at 970dc8c> python scripts/dtype_parity.py <checkpoint dir>   # once per checkpoint
```

The end-to-end case, from a laya checkout:

```python
import sys, torch, laya
sys.path.insert(0, "benchmarks"); import parity_fast as pf
a = laya.load("convaiinnovations/laya")                  # cc >= 8: a.dtype == torch.bfloat16
t = pf.TEXTS[1]; state = {"subject": t[:60], "body": t}; qs = pf.PRESETS["moderation"]()
print(a.predict(state, qs)["answers"]["spam"]["noul"])   # 0.505
a.dtype = torch.float16
print(a.predict(state, qs)["answers"]["spam"]["noul"])   # 0.469
```

## Laya version / Python / OS / Device

0.3.20 (`main` @ 970dc8c), Python 3.12.3, torch 2.14.0+cu130, transformers 5.17.0, Ubuntu 24.04, NVIDIA RTX 2000 Ada Generation Laptop GPU (cc 8.9), driver 590.48.01. `TORCH_DISABLE_NATIVE_JIT=1` was set (#365 / #436); it does not change these numbers.

## Suggestion

1. A `LAYA_CUDA_AMP=fp16|bf16` override mirroring `LAYA_CPU_AMP`, plus a README sentence under Confidence Gating naming the default and these numbers, so whoever fits a threshold knows which dtype they are fitting it in. PR coming alongside this issue.
2. Your call: whether the cc >= 8 default should be fp16 rather than the checkpoint's training dtype. It moves every CUDA user's probabilities (towards fp32, by up to 0.07), so it is a numbers-moving change I would rather not decide inside a PR. If you want it, the same script gives the before/after for BENCHMARKS.md. The TileLang fast path (`laya/fast.py`, bf16-resident weights) is a separate forward and is not affected either way.
